common/js/src/executable-module/offscreen-doc-emscripten-module.js

```diff
@@ -0,0 +1,256 @@
+  /**
+   * @param {string} moduleName
+   */
+  constructor(moduleName) {
+    super();
+    /** @type {string} @const @private */
+    this.moduleName_ = moduleName;
+    /** @type {!goog.log.Logger} @const @private */
+    this.logger_ = GSC.Logging.getScopedLogger('OffscreenDocEmscriptenModule');
+    /** @type {boolean} @private */
+    this.offscreenDocActive_ = false;
+    /** @type {!OffscreenDocMessageChannel} @const @private */
+    this.messageChannel_ = new OffscreenDocMessageChannel();
+    /** @type {!GSC.PortMessageChannel|null} @private */
+    this.statusChannel_ = null;
+    /** @type {!goog.promise.Resolver<void>} @const @private */
+    this.loadPromiseResolver_ = goog.Promise.withResolver();
+    GSC.PromiseHelpers.suppressUnhandledRejectionError(
+        this.loadPromiseResolver_.promise);
+  }
+
+  /** @override */
+  getLogger() {
+    return this.logger_;
+  }
+
+  /** @override */
+  startLoading() {
+    // Listen for incoming message channels from the Offscreen Document once
+    // it's loaded.
+    chrome.runtime.onConnect.addListener((port) => this.onConnect_(port));
+
+    // Load the Offscreen Document. The loading arguments are passed via URL
+    // query parameters.
+    const url = new URL(OFFSCREEN_DOC_URL, globalThis.location.href);
+    url.searchParams.append(URL_PARAM, this.moduleName_);
+    chrome.offscreen.createDocument(
+        {
+          'url': url.toString(),
+          'reasons': OFFSCREEN_DOC_REASONS,
+          'justification': OFFSCREEN_DOC_JUSTIFICATION,
+        },
+        () => this.onOffscreenDocCreated_());
+    // Once the document is loaded, it'll immediately start loading the
+    // Emscripten module and report the status to us via m
+  }
+
+  /** @override */
+  getLoadPromise() {
+    return this.loadPromiseResolver_.promise;
+  }
+
+  /** @override */
+  getMessageChannel() {
+    return this.messageChannel_;
+  }
+
+  /** @override */
+  disposeInternal() {
+    if (this.offscreenDocActive_)
+      chrome.offscreen.closeDocument();
+    this.messageChannel_.dispose();
+    if (this.statusChannel_)
+      this.statusChannel_.dispose();
+    this.loadPromiseResolver_.reject();
+    super.disposeInternal();
+  }
+
+  /** @private */
+  onOffscreenDocCreated_() {
+    if (chrome && chrome.runtime && chrome.runtime.lastError) {
+      const error =
+          goog.object.get(chrome.runtime.lastError, 'message', 'Unknown error');
+      goog.log.warning(
+          this.logger_, `Failed to load the offscreen document: ${error}`);
+    } else {
+      this.offscreenDocActive_ = true;
+    }
+
+    if (this.offscreenDocActive_ && this.isDisposed()) {
+      // Edge case: we got destroyed while waiting for the
+      // `chrome.offscreen.createDocument()` completion; make sure to not leave
+      // the doc in that case as well.
+      chrome.offscreen.closeDocument();
+    } else if (!this.offscreenDocActive_) {
+      this.dispose();
```

Comment:
Unless this function is called multiple times, I think this line can be put after line 141

Reply:
Ack - this chunk of code doesn't exist anymore after the rewrite.